README.md
# NAME

DBIx::RecordFactory - It's new $module

# SYNOPSIS

    use DBIx::RecordFactory;
    my $factory = DBIx::RecordFactory->new(teng => $teng);
    $factory->define('user' => +{
        id => sub { shift->sequence('account_id') },
        name => sub { shift->choice(qw(foo bar baz)); },
        account_id => sub {
            my $account = shift->insert('account');
            $account->{id};
        }
    });
    $factory->define('account' => {
        login_id => sub { shift->string(255) },
        password => sub { shift->string(255) },
    });

    my $userdata = $factory->insert('user');
    #   => {
    #     'id' => 1,
    #     'account_id' => 1,
    #   }
    my $teng = Teng->new(dbh => $dbh);
    my $account = $teng->single(account => { id => $userdata->{account_id} });

# DESCRIPTION

DBIx::RecordFactory is ...

# SEE ALSO

    FactoryGirl - https://github.com/thoughtbot/factory_girl

# LICENSE

Copyright (C) Keiji, Yoshimi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Keiji, Yoshimi <walf443@gmail.com>


Changes
Revision history for Perl extension DBIx-RecordFactory

{{$NEXT}}

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'DBIx-RecordFactory',
    module_name     => 'DBIx::RecordFactory',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/DBIx/RecordFactory/RuleSet/ActiveRecord.pm
package DBIx::RecordFactory::RuleSet::ActiveRecord;
use 5.008_001;
use strict;
use warnings;
use DBI qw(:sql_types);

our $VERSION = '0.01';

sub apply {
    my ($class, %args) = @_;
    my $factory = $args{factory};

    my $schema = $factory->teng->schema;
    for my $table_name ( keys %{ $schema->{tables} } ) {
        my $table = $schema->get_table($table_name);
        my $columns = {};
        my $primary_keys = $table->primary_keys;
        my $primary_key;
        if ( @$primary_keys == 1 ) {
            $primary_key = $primary_keys->[0];
        } else {
        }
        for my $col ( @{ $table->columns } ) {
            if ( $primary_key && $col eq $primary_key ) {
                $columns->{$col} = sub { $_[0]->sequence($table_name . '.' . $col) };
            } elsif ( $col =~ /^(.+)_id$/ ) {
                my $relation_table = $1;
                if ( defined $args{$relation_table} && ref $args{$relation_table} eq "HASH" ) {
                    $columns->{$col} = sub { my $relation = $_[0]->insert($relation_table, %{$args{$relation_table}}); return $relation->{id} };
                } else {
                    $columns->{$col} = sub { my $relation = $_[0]->insert($relation_table); return $relation->{id} };
                }
            } else {
                my $type = $table->get_sql_type($col);
                if ( $type == SQL_INTEGER ) {
                    $columns->{$col} = sub { $_[0]->uint(1000) };
                } elsif ( $type == SQL_VARCHAR ) {
                    $columns->{$col} = sub { $_[0]->string(10, 20) };
                } else {
                    # does not support. you should define manually,
                }
            }
        }
        $factory->define($table_name => $columns);
    }
}


1;
__END__

=head1 NAME

DBIx::RecordFactory::RuleSet::ActiveRecord - Perl extension to do something

=head1 VERSION

This document describes DBIx::RecordFactory::RuleSet::ActiveRecord version 0.01.

=head1 SYNOPSIS

    use DBIx::RecordFactory::RuleSet::ActiveRecord;
    my $factory = DBIx::RecordFactory->new(dbh => $dbh);
    DBIx::RecordFactory::RuleSet::ActiveRecord->apply(factory => $factory);
    $factory->redefine(xxx => { yyy_id => sub { my $r = $_[0]->insert('yyy'); $r->{id} } });

=head1 DESCRIPTION


=head1 INTERFACE

=head2 Functions

=head3 C<< hello() >>

# TODO

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

<<YOUR NAME HERE>> E<lt><<YOUR EMAIL ADDRESS HERE>>E<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2013, <<YOUR NAME HERE>>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/DBIx/RecordFactory.pm
package DBIx::RecordFactory;
use 5.008005;
use strict;
use warnings;
use Class::Accessor::Lite ( new => 1, ro => [qw( teng )] );
use Teng::Schema::Loader;
use POSIX qw();
use DBI;

our $VERSION = "0.01";

sub dbh {
    $_[0]->teng->dbh;
}

sub define {
    my ($self, $name, $code_hash, %args) = @_;
    $self->{__rule} ||= {};
    $self->{__rule}->{ $name } = $code_hash;
    $self->{__table_of} ||= {};
    $self->{__table_of}->{ $name } = ( $args{'table'} || $name );

    return $code_hash;
}

sub redefine {
    my ($self, $name, $code_hash, %args) = @_;
    my $rule = $self->rule($name)
        or die "Can't find rule: $name";

    my $new_rule = {
        %$rule,
        %$code_hash,
    };
    $self->{__table_of}->{ $name } = $new_rule;
    $self->{__table_of} ||= {};
    $self->{__table_of}->{ $name } = ( $args{'table'} || $name );

    return $new_rule;
}

sub table_name {
    my ($self, $name) = @_;
    return $self->{__table_of}->{ $name };
}

sub rule {
    my ($self, $name) = @_;
    return $self->{__rule}->{$name};
}

sub insert {
    my ($self, $rule, %args) = @_;
    my $table = $self->table_name($rule);
    my $data = $self->build($rule, %args);
    my $original_suppress_row_object = $self->teng->suppress_row_objects;
    $self->teng->suppress_row_objects(1);
    my $result = $self->teng->insert($table => $data);
    $self->teng->suppress_row_objects($original_suppress_row_object);
    return $result;
}

sub build {
    my ($self, $rule_name, %args) = @_;
    my $rule = $self->rule($rule_name)
        or die "Can't find rule: $rule_name";

    my $row = {};
    for my $col ( keys %$rule ) {
        if ( $args{$col} ) {
            if ( ref $args{$col} && ref $args{$col} eq "CODE" ) {
                $row->{$col} = $args{$col}->($self);
            } else {
                $row->{$col} = $args{$col};
            }
        } else {
            $row->{$col} = $rule->{$col}->($self);
        }
    }
    return $row;
}

sub sequence {
    my ($self, $namespace) = @_;
    $self->{__sequence} ||= {};
    $self->{__sequence}->{$namespace} ||= 0;
    $self->{__sequence}->{$namespace}++;
    return $self->{__sequence}->{$namespace};
}

sub string {
    my ($self, $max, $min) = @_;
    $min ||= 0;
    my $count = $min + POSIX::floor(rand() * ($max - $min));
    my @alpha = qw(a b c d e f g h i j k l m n o p q r s t u v x y z 1 2 3 4 5 6 7 8 9 0 _ );
    my $result = "";
    for my $loop (1..$count) {
        my $index = POSIX::floor(rand() * scalar @alpha);
        my $choice = $alpha[$index];
        $result .= $choice;
    }
    return $result;
}

sub uint {
    my ($self, $max, $min) = @_;
    $min ||= 0;
    return $min + POSIX::floor(rand() * ($max - $min));
}

sub choice {
    my ($self, @args) = @_;
    $args[$self->uint(@args - 1)];
}

1;
__END__

=encoding utf-8

=head1 NAME

DBIx::RecordFactory - It's new $module

=head1 SYNOPSIS

    use DBIx::RecordFactory;
    my $factory = DBIx::RecordFactory->new(teng => $teng);
    $factory->define('user' => +{
        id => sub { shift->sequence('account_id') },
        name => sub { shift->choice(qw(foo bar baz)); },
        account_id => sub {
            my $account = shift->insert('account');
            $account->{id};
        }
    });
    $factory->define('account' => {
        login_id => sub { shift->string(255) },
        password => sub { shift->string(255) },
    });

    my $userdata = $factory->insert('user');
    #   => {
    #     'id' => 1,
    #     'account_id' => 1,
    #   }
    my $account = $factory->teng->single(account => { id => $userdata->{account_id} });

=head1 DESCRIPTION

DBIx::RecordFactory is ...

=head1 SEE ALSO

 FactoryGirl - https://github.com/thoughtbot/factory_girl

=head1 LICENSE

Copyright (C) Keiji, Yoshimi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Keiji, Yoshimi E<lt>walf443@gmail.comE<gt>

=cut

